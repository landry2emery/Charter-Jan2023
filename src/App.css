.Container {
    display: flex;
    justify-content: center;
    align-items: flex-start;
    flex-wrap: wrap;
    width: 100vw;
    height: 100vh;
}

.Content {
    display: flex;
    align-items: flex-start;
    flex-wrap: wrap;
    width: 80%;
}


.Data {
    width: 100%;
    height: 400px;
    overflow:scroll;
}

.Data ul li {
    display: flex;
    justify-content: space-between;
    padding: 10px;
}

.Data ul li span {
    width: 30%;
}

li {
    list-style-type: none;
    padding: 5px;
}

.Button{
    width: 100%;
}

button{
    padding: 8px;
    border: 1px solid #008CBA;;
    border-radius: 10px;
    margin-left: 10px;
}

.ShowResults {
    display: flex;
    flex-direction: column;
    height: 400px;
    overflow: scroll;
    width: 100%;
    padding: 5px;
}

.ShowResults div{
    padding: 5px;
}


