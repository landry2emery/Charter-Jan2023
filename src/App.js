import { useEffect, useState } from "react";
import axios from "axios";
import "./App.css";

import { DisplayPoints } from "./functions/DisplayPoints.js";
import { GetAvailableMonths } from "./functions/GetAvailableMonths.js";

const URL = "https://63c5df26f3a73b34785f0b2d.mockapi.io/records";

function App() {
  const [data, setData] = useState([]);
  const [pointsRecord, setPointsRecord] = useState({});

  useEffect(() => {
    axios.get(URL).then((records) => {
      setData(records.data);
    });
  }, []);

  const handleOnButtonClick = () => {
    setPointsRecord(DisplayPoints(data));
  };

  return (
    <div className="Container">
      <div className="Content">
        <h1>Customer Rewards</h1>
        <div className="Data">
          <ul >
            {data.length > 0 ? (
              <>
                {data.map((customer) => {
                  return (
                    <li data-testid="customer" key={customer.id}>
                      <span> {customer.name} </span>
                      <span> {customer.date} </span>
                      <span> {customer.amount} </span>
                    </li>
                  );
                })}
              </>
            ) : (
              <></>
            )}
          </ul>
        </div>
        <div className="Button">
          <button onClick={handleOnButtonClick}>Display Points</button>
        </div>
        <div className="ShowResults">
          {Object.keys(pointsRecord).length > 0 ? (
            <ul>
              {Object.values(pointsRecord).map((record) => {
                return (
                  <li data-testid={`point-${record.name}`} key={record.name}>
                    {record.name} has obtained in{" "}
                    {console.log(record.pointsPerMonth)}
                    {GetAvailableMonths(record.pointsPerMonth)} and the total
                    points is <strong>{record.total}</strong>{" "}
                  </li>
                );
              })}
            </ul>
          ) : (
            <></>
          )}
        </div>
      </div>
    </div>
  );
}

export default App;
