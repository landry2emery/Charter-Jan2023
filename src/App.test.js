import {
  cleanup,
  render,
  waitFor,
  scree,
  getAllByTestId,
  screen,
  fireEvent,
  findAllByTestId,
} from "@testing-library/react";
import "@testing-library/jest-dom";
import App from "./App";
import axios from "axios";

describe("Testing Api and Points calculation", () => {
  let records;
  let response;
  beforeEach(() => {
    records = [
      { id: "1", name: "Julie Navarro", date: "12/16/2022", amount: 120 },
      { id: "2", name: "Mia Swanson", date: "1/7/2023", amount: 92.3 },
      { id: "3", name: "Hayden Robbins", date: "2/27/2023", amount: 36.95 },
      { id: "4", name: "Aurora Gross", date: "12/13/2022", amount: 55.05 },
      { id: "5", name: "Aurora Gross", date: "12/18/2022", amount: 160.15 },
      { id: "6", name: "Aurora Gross", date: "2/9/2022", amount: 93 },
    ];
    response = {
      data: records,
      status: 200,
      statusText: "OK",
      headers: {},
      config: {},
    };
    jest.mock("axios");
    axios.get = jest.fn();
    axios.get.mockResolvedValue(response);
  });
  //afterEach(cleanup);
  test("/renders App correctly", async () => {
    const { asFragment, findByText } = render(<App />);
    const customerRewards = await findByText("Customer Rewards")
    expect(customerRewards).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test("/Api call and Points calculation works", async () => {
    const {asFragment,findByText } = render(<App />);
    const IdCustomer = await screen.findAllByTestId("customer");
    expect(IdCustomer).toHaveLength(6);

    const button = screen.getByRole("button")

    fireEvent.click(button);
    expect(button).toHaveTextContent("Display Points");
    
    const IdJulia = await screen.findByTestId("point-Julie Navarro");
    expect(IdJulia).toBeInTheDocument();
    expect(IdJulia).toHaveTextContent("Julie Navarro has obtained in December: 90 Points, and the total points is 90");

    const IdMia = await screen.findByTestId("point-Mia Swanson");
    expect(IdMia).toBeInTheDocument();
    expect(IdMia).toHaveTextContent("Mia Swanson has obtained in January: 42 Points, and the total points is 42");

    expect(async() => {
      const IdHayden = await screen.findByTestId("point-Hayden Robbins");
      IdHayden.toThrow("Unable to find an element")
    })

    const IdAurora = await screen.findByTestId("point-Aurora Gross");
    expect(IdAurora).toBeInTheDocument();
    expect(IdAurora).toHaveTextContent("Aurora Gross has obtained in February: 43 Points, December: 175 Points, and the total points is 218");

    // expect(IdHayden).not.toBeInTheDocument();
    //.toThrow("Unable to find an element")

  });

});
