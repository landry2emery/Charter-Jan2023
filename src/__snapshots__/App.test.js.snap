// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Api and Points calculation /renders App correctly 1`] = `
<DocumentFragment>
  <div
    class="Container"
  >
    <div
      class="Content"
    >
      <h1>
        Customer Rewards
      </h1>
      <div
        class="Data"
      >
        <ul />
      </div>
      <div
        class="Button"
      >
        <button>
          Display Points
        </button>
      </div>
      <div
        class="ShowResults"
      />
    </div>
  </div>
</DocumentFragment>
`;
